File: android/app/src/main/java/com/dailysns/viewmodels/AlbumViewModel.kt
package com.dailysns.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dailysns.models.Album
import com.dailysns.services.DataService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * AlbumViewModel.kt
 * DailySNSWithYourSS Android版
 * 
 * アルバム閲覧のビジネスロジック
 */

class AlbumViewModel(
    private val dataService: DataService = DataService()
) : ViewModel() {
    
    private val _albums = MutableStateFlow<List<Album>>(emptyList())
    val albums: StateFlow<List<Album>> = _albums.asStateFlow()
    
    private val _selectedAlbum = MutableStateFlow<Album?>(null)
    val selectedAlbum: StateFlow<Album?> = _selectedAlbum.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    /**
     * アルバムを取得
     */
    fun fetchAlbums() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                _albums.value = dataService.fetchAlbums()
            } catch (e: Exception) {
                _errorMessage.value = e.message
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    /**
     * アルバムを選択
     */
    fun selectAlbum(album: Album) {
        _selectedAlbum.value = album
    }
    
    /**
     * 選択をクリア
     */
    fun clearSelection() {
        _selectedAlbum.value = null
    }
}

File: web/src/viewmodels/AlbumViewModel.ts
/**
 * AlbumViewModel.ts
 * DailySNSWithYourSS Web版
 * 
 * アルバム閲覧のビジネスロジック
 */

import { Album } from '../models/User';
import { DataService } from '../services/DataService';

export class AlbumViewModel {
  private _albums: Album[] = [];
  private _selectedAlbum: Album | null = null;
  private _isLoading: boolean = false;
  private _errorMessage: string | null = null;
  private listeners: Set<() => void> = new Set();
  private dataService: DataService;

  constructor(dataService: DataService = new DataService()) {
    this.dataService = dataService;
  }

  get albums(): Album[] {
    return this._albums;
  }

  get selectedAlbum(): Album | null {
    return this._selectedAlbum;
  }

  get isLoading(): boolean {
    return this._isLoading;
  }

  get errorMessage(): string | null {
    return this._errorMessage;
  }

  /**
   * 変更を監視
   */
  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notify(): void {
    this.listeners.forEach(listener => listener());
  }

  /**
   * アルバムを取得
   */
  async fetchAlbums(): Promise<void> {
    this._isLoading = true;
    this._errorMessage = null;
    this.notify();

    try {
      this._albums = await this.dataService.fetchAlbums();
    } catch (error) {
      this._errorMessage = error instanceof Error ? error.message : 'Unknown error';
    } finally {
      this._isLoading = false;
      this.notify();
    }
  }

  /**
   * アルバムを選択
   */
  selectAlbum(album: Album): void {
    this._selectedAlbum = album;
    this.notify();
  }

  /**
   * 選択をクリア
   */
  clearSelection(): void {
    this._selectedAlbum = null;
    this.notify();
  }
}
